### PR TITLE
feat: Add SFT (Supervised Fine-Tuning) support to Qwen3ForCausalLM training

### PR DESCRIPTION
**Pull request by canqin001**
## Overview
This PR adds comprehensive SFT (Supervised Fine-Tuning) support to the Qwen3ForCausalLM training infrastructure, enabling instruction-following training while maintaining backward compatibility with existing pre-training workflows.

## Key Features
- **Dual Training Modes**: Support for both pre-training (all tokens masked) and SFT (only response tokens masked)
- **Multiple Data Formats**: Native support for Alpaca, ShareGPT, and custom instruction-response formats
- **Backward Compatibility**: Existing pre-training code works unchanged
- **Automatic Masking**: Intelligent src_mask generation for instruction preservation during SFT

### Model Extensions
- Extended `Qwen3ForCausalLM.forward()` with SFT parameters (`src_mask`, `training_mode`)
- Added `create_sft_src_mask()` helper method
- Simplified conditional logic to avoid XLA compilation issues

### Data Processing
- **New**: `SFTDataCollator` for handling instruction-response pairs
- **New**: `create_sft_dataset()` for processing raw datasets
- Support for system prompts, custom separators, and multiple formats

### Training Infrastructure
- Modified `train.py` to support SFT mode via `training_mode` parameter
- Added batch validation for SFT training
- Updated data loading pipeline for instruction-response datasets

### Configuration
- **New**: `configs/data/sft.yaml` for SFT-specific data configuration
- Added `training_mode` parameter to default config
- Extended attention module to support "default" kernel for SFT stability

### Documentation
- **New**: Comprehensive SFT training guide (`SFT_TRAINING_README.md`)
- Updated training scripts with SFT examples

### SFT Training
```bash
./recipes/train_qwen3_1.7b_sft.sh
```

### Pre-training (unchanged)
```bash
./recipes/train_qwen3_1.7b.sh
```

### Alpaca Format
```json
{
  "instruction": "What is the capital of France?",
  "output": "The capital of France is Paris.",
  "system": "You are a helpful assistant."
}
```

### ShareGPT Format
```json
{
  "conversations": [
    {"from": "human", "value": "What is the capital of France?"},
    {"from": "assistant", "value": "The capital of France is Paris."}
  ]
}
```

## Testing
- Verified backward compatibility with existing pre-training workflows
- Tested with SFT dataset formats (Alpaca Only, ShareGPT and custom not yet)
- Validated XLA compilation and tensor shape handling
- Confirmed proper instruction-response separation and masking

## Files Changed
- `torchprime/torch_xla_models/flex/modeling_qwen.py`
- `torchprime/torch_xla_models/train.py`
- `torchprime/torch_xla_models/flex/attention.py`
- `torchprime/torch_xla_models/sft_data_collator.py` (new)
- `torchprime/torch_xla_models/configs/default.yaml`
- `torchprime/torch_xla_models/configs/data/sft.yaml` (new)
- `recipes/train_qwen3_1.7b_sft.sh` (new)
- `torchprime/torch_xla_models/SFT_TRAINING_README.md` (new)

### CODE REVIEW

File: torchprime/torch_xla_models/flex/attention.py [hunk 1]
```diff
@@ -86,15 +86,15 @@ def forward(
       case "flash_attention":
         # Integrated with PyTorch/XLA Pallas Flash Attention:
         default_block_sizes = {
-          "block_q": 2048,
+          "block_q": 1024,
```
Comment (hlnchen) [inline torchprime/torch_xla_models/flex/attention.py:89]: Why do we change this?

File: torchprime/torch_xla_models/flex/modeling_qwen.py [hunk 2]
```diff
@@ -367,7 +367,12 @@ def forward(
     causal_mask = causal_mask.unsqueeze(0).unsqueeze(0)  # Add batch and head dimension
 
     if attention_mask is not None:
-      causal_mask = causal_mask * attention_mask[:, None, None, :]
+      # Ensure attention_mask has valid dimensions
+      if attention_mask.dim() == 2:
```
Comment (hlnchen) [inline torchprime/torch_xla_models/flex/modeling_qwen.py:371]: If this is cursor-generated then it could be not necessary. We actually don't use attention mask in training, especially for flash attention we set attention to be bidirectional.

File: torchprime/torch_xla_models/flex/modeling_qwen.py [hunk 3]
```diff
@@ -409,12 +414,40 @@ def _init_weights(self, module):
       if module.padding_idx is not None:
         module.weight.data[module.padding_idx].zero_()
 
+  def create_sft_src_mask(
```
Comment (hlnchen) [inline torchprime/torch_xla_models/flex/modeling_qwen.py:417]: where is this function used?

File: torchprime/torch_xla_models/train.py [hunk 4]
```diff
@@ -628,32 +706,103 @@ def main(config: DictConfig):
...
+          names=config.data.gcs_dataset_names,
+          weights=config.data.weights,
+          tokenizer=tokenizer,
+          seed=config.seed,
+          block_size=config.data.block_size,
+        )
+      )
+    else:
+      raise ValueError("No dataset provided")
+  # data = split_dataset_by_node(data, xr.process_index(), xr.process_count()) # not working as expected, will only load a subset of the dataset
```
Comment (hlnchen) [inline torchprime/torch_xla_models/train.py:798]: I think we still need to uncomment this line 😂

File: remote_run_sft.sh [hunk 5]
```diff
@@ -1,7 +1,8 @@
-TPU_VM_NAME="<your_tpu_vm_name>"
+TPU_VM_NAME="sfr-cqin-v4-16"
 TPU_ZONE="us-central2-b"
-BRANCH="<your_branch>"
-RECIPE="recipes/train_qwen3_1.7b.sh"
+BRANCH="cqin/dev"
+# RECIPE="recipes/train_qwen3_1.7b.sh"
+RECIPE="recipes/train_qwen3_1.7b_sft.sh"
```
[the remark this answers is not in the document]
Comment (canqin001) [reply remote_run_sft.sh:5]: yeah, this script is redundant

File: torchprime/torch_xla_models/flex/modeling_qwen.py (hunk 2 above)
Comment (canqin001) [reply torchprime/torch_xla_models/flex/modeling_qwen.py:371]: this is generated during my debug of sft code, maybe it is innecessary

File: torchprime/torch_xla_models/flex/modeling_qwen.py (hunk 3 above)
Comment (canqin001) [reply torchprime/torch_xla_models/flex/modeling_qwen.py:417]: yeah, this would be redundant, let me check

File: copy_env.sh [hunk 6]
```diff
@@ -1,4 +1,4 @@
-TPU_VM_NAME="<your_tpu_vm_name>"
+TPU_VM_NAME="sfr-cqin-v4-16"
```
[the remark this answers is not in the document]
Comment (canqin001) [reply copy_env.sh:1]: yeah, i would revise it

**Review by hlnchen — APPROVED** [review]
LGTM
